feat: jwt 기반 인증기능 구현

taeung515 opened the pull request:
## 이슈 번호
#32 

## 작업 내용
- JWT 유틸클래스 생성
- JWT 필터 구현

## 스크린샷(선택)

taeung515 committed:
feat: 에러 응답 포맷 및 JWT 시크릿키 설정 추가

- application.yml에 커스텀 에러 응답 포맷 추가
- JWT 시크릿 키 및 관련 설정 추가

resolves: #32

taeung515 committed:
feat: JWT 의존성 추가

resolves: #32

taeung515 committed:
feat: JWT 유틸 클래스 추가

resolves: #32

taeung515 committed:
feat: 로그인 요청 dto 생성

resolves: #32

taeung515 committed:
feat: HttpRequest에 저장할 key 값 상수 클래스 추가

resolves: #32

taeung515 committed:
feat: 로그인유저 이메일 검증 로직 추가

resolves: #32

taeung515 committed:
feat: 로그인 기능 구현 및 JWT 토큰 발급

- 이메일 및 비밀번호 기반 로그인 처리

resolves: #32

GyeongSe99 commented on newsfeed/src/main/java/com/npcamp/newsfeed/auth/controller/AuthController.java:
```diff
@@ -41,4 +52,28 @@ public ResponseEntity<ApiResponse<Void>> deleteUser(
+    @PostMapping("/login")
+    public ResponseEntity<ApiResponse<Void>> login(@RequestBody LoginRequestDto requestDto) {
+
+        String email = requestDto.getEmail();
+        String password = requestDto.getPassword();
+
+        // 해당 이메일 유저 조회
+        User user = userRepository.getUserByEmailOrElseThrow(email);
+
+        // 비밀번호 일치여부 확인
+        if (!encoder.matches(password, user.getPassword())) {
+            throw new ResourceForbiddenException(ErrorCode.INVALID_PASSWORD);
+        }
+
+        // JWT 토큰 발급 (userId 기반)
+        String token = jwtUtil.generateToken(user.getId());
```
비지니스 로직은 Service에 구현해주세요

taeung515 replied:
네 알겠습니다! Service계층에 구현하도록 하겠습니다.

GyeongSe99 commented on newsfeed/src/main/java/com/npcamp/newsfeed/common/security/JwtUtil.java:
```diff
@@ -0,0 +1,107 @@
+@Component
+public class JwtUtil {
+
+    public static final String BEARER_PREFIX = "Bearer ";
+    // JWT 토큰의 만료 시간
+    private final long TOKEN_TIME = 60 * 60 * 1000L;
```
상수는 @value를 주입하여 사용하고 
실제값은 application.yml에 작성하여  따로 관리해주는 것이 좋을 것 같습니다.

GyeongSe99 replied:
특히 TOKEN_TIME같이 정책에 따라 자주 바뀔 수 있는 부분은 설정 파일에서 관리하는 것이 찾기가 쉬워 더 좋습니다.

taeung515 replied:
네 알겠습니다! 설정파일에서 관리하도록 수정하겠습니다!

GyeongSe99 commented on newsfeed/src/main/java/com/npcamp/newsfeed/common/constant/RequestAttribute.java:
```diff
@@ -4,8 +4,9 @@
 
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
-public class SessionKeys {
+public class RequestAttribute {
 
-    public static final String SESSION_USER_ID = "loginUserId";
+    // 토큰에서 꺼내온 Value 를 저장할 때 쓸 Key
+    public static final String USER_ID = "userId";
```
해당 부분 계속 상수로 쓰이는 것이라면 Enum으로 관리하는 것이 더 좋을 것 같습니다.

taeung515 replied:
Enum으로 쓴다면 어떤 장점이 있을까요?

GyeongSe99 replied:
태웅님이 사용하신 방법이 "클래스 상수"로 구현한 방법입니다. 
예전에는 final static을 사용하여 상수를 관리하였는데 상수를 좀 더 안전하게 관리하기 위해 Enum이 생기게 되었습니다.

Enum을 사용했을 때의 장점으로는
- 타입 안전: Enum하나하나가 클래스 형태로 취급하기 때문에 잘못된 키 사용 시 컴파일 에러로 방지 가능
- 자동완성, 리팩토링 등 사용가능
- 각 키에 의미 있는 메서드나 속성을 부여할 수 있음(String값이라던가 int, boolean나 ErrorCode의 status등)
  - 그만큼 확장성도 있겠죠?
- 그룹화와 구조화가 명확하다.
위의 장점 때문에 Enum을 사용하는 것이 유지보수성이 더 뛰어나다고 생각합니다 ^-^

GyeongSe99 replied:
https://inpa.tistory.com/entry/JAVA-%E2%98%95-%EC%97%B4%EA%B1%B0%ED%98%95Enum-%ED%83%80%EC%9E%85-%EB%AC%B8%EB%B2%95-%ED%99%9C%EC%9A%A9-%EC%A0%95%EB%A6%AC#enum%EC%9D%98_%EC%9E%A5%EC%A0%90

관련 블로그 글인데 제가 몰랐던 장점도 있어서 참고하시면 좋을 것 같아요

taeung515 replied:
enum의 장점을 명확히 알 수 있는 좋은 글 공유 감사합니다. 하지만, 이번 프로젝트에서는 Spring Security를 사용하지 않기 때문에, JWT 토큰에서 추출한 userId 값을 HttpServletRequest에 저장하고, 이후 컨트롤러에서 @RequestAttribute(key)를 통해 해당 값을 꺼내는 방식으로 처리합니다.
@RequestAttribute는 String 타입의 key만 지원하므로, enum을 사용하는 방식은 적용하기 어렵고, 문자열 상수를 사용해야 할 것 같습니다!

taeung515 committed:
feat: 비지니스 로직 컨트롤러계층 -> 서비스계층 위임

resolves: #32

taeung515 committed:
feat: 상수 설정파일에서 관리

resolves: #32

taeung515 committed:
chore: 쓰지않는 import 코드포맷팅

resolves: #32

taeung515 committed:
chore: 상수 static으로 변경

resolves: #32

taeung515 committed:
chore: generateToken -> login으로 메서드 이름 변경

- 메서드 의도를 명확히 하기 위함

resolves: #32

taeung515 committed:
refactor: private -> public 상수 접근제어자 변경

resolves: #32

taeung515 committed:
chore: 서비스계층 login 로직에 주석 추가

resolves: #32

taeung515 committed:
refactor: 필터 내 예외 응답 형식 통일

- 필터에서 발생한 예외는 @ControllerAdvice로 처리되지 않으므로, 별도의 예외 처리 메서드를 추출하여 모든 응답을 ApiResponse형태로 통일함.

resolves: #32

taeung515 committed:
chore: 코드 포맷팅

resolves: #32

taeung515 committed:
refactor: deleted 필드에서 final 제거

- 논리 삭제 처리 시 값 변경을 위해 final 키워드를 제거

resolves: #32

taeung515 committed:
Merge branch 'dev' into feat/jwt

# Conflicts:
#	newsfeed/src/main/java/com/npcamp/newsfeed/follow/controller/FollowController.java
#	newsfeed/src/main/java/com/npcamp/newsfeed/post/controller/PostController.java
#	newsfeed/src/main/java/com/npcamp/newsfeed/post/service/PostService.java

taeung515 committed:
refactor: LoginRequestDto email 필드에 final 추가

resolves: #32

taeung515 committed:
refactor: 레포지토리에 구현하는 메서드 이름 변경

- JPA 쿼리 메서드와 네이밍 통일

resolves: #32
